include/nbl/builtin/hlsl/emulated/emulated_float64_t_utils.hlsl

```diff
@@ -0,0 +1,459 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_UTILS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_UTILS_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/type_traits.hlsl>
+#include <nbl/builtin/hlsl/emulated/emulated_float64_t.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+// TODO: enable
+//template<typename device_capabilities = void, bool FastMath = false, bool FlushDenormToZero = true>
+//using portable_float64_t = conditional_t<device_capability_traits<device_capabilities>::shaderFloat64, float64_t, typename emulated_float64_t<FastMath, FlushDenormToZero> >;
+
+#ifndef __HLSL_VERSION
+template<typename device_capabilities = void, bool FastMathIfEmulated = false, bool FlushDenormToZeroIfEmulated = true>
+using portable_float64_t = typename conditional<true, float64_t, emulated_float64_t<FastMathIfEmulated, FlushDenormToZeroIfEmulated> >::type;
+#else
+template<typename device_capabilities = void, bool FastMathIfEmulated = false, bool FlushDenormToZeroIfEmulated = true>
+using portable_float64_t = typename conditional<true, float64_t, emulated_float64_t<FastMathIfEmulated, FlushDenormToZeroIfEmulated> >::type;
+#endif
+
+template<typename EmulatedType, uint32_t N>
+struct emulated_vector {};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 2>
+{
+    using type = emulated_vector<EmulatedType, 2>;
+
+    EmulatedType x;
+    EmulatedType y;
+
+    EmulatedType calcComponentSum() NBL_CONST_MEMBER_FUNC
+    {
+        return x + y;
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE type create(EmulatedType x, EmulatedType y)
+    {
+        type output;
+        output.x = x;
+        output.y = y;
+
+        return output;
+    }
+
+    type operator+(float rhs)
+    {
+        type output;
+        EmulatedType rhsAsEF64 = EmulatedType::create(rhs);
+        output.x = x + rhsAsEF64;
+        output.y = y + rhsAsEF64;
+
+        return output;
+    }
+
+    type operator+(EmulatedType rhs)
+    {
+        type output;
+        output.x = x + rhs;
+        output.y = y + rhs;
+
+        return output;
+    }
+
+    type operator+(type rhs)
+    {
+        type output;
+        output.x = x + rhs.x;
+        output.y = y + rhs.y;
+
+        return output;
+    }
+
+    type operator-(float rhs)
+    {
+        return create(x, y) + (-rhs);
+    }
+
+    type operator-(EmulatedType rhs)
+    {
+        return create(x, y) + (rhs.flipSign());
+    }
+
+    type operator-(type rhs)
+    {
+        rhs.x = rhs.x.flipSign();
+        rhs.y = rhs.y.flipSign();
```

Comment:
don't hardcode your `emulated_vector` to only work with IEEE754 base scalar types